File: src/Flux.Web/src/app/models/subscription.ts
export enum SubscriptionCategory {
  Entertainment = 0,
  Insurance = 1,
  Utilities = 2,
  Mobile = 3,
  Internet = 4,
  Productivity = 5,
  Health = 6,
  Education = 7,
  Transportation = 8,
  Other = 9
}

export enum SubscriptionBillingCycle {
  Weekly = 0,
  Monthly = 1,
  Quarterly = 2,
  Yearly = 3
}

export enum SubscriptionStatus {
  Active = 0,
  Paused = 1,
  Cancelled = 2
}

export interface Subscription {
  id: number;
  ownerUserId: string;
  ownerUsername: string;
  serviceName: string;
  providerName: string;
  category: SubscriptionCategory;
  tags: string[];
  billingCycle: SubscriptionBillingCycle;
  amount: number;
  currencyCode: string;
  startDateUtc: string;
  nextDueDateUtc: string;
  reminderDaysBeforeDue: number;
  autoRenew: boolean;
  status: SubscriptionStatus;
  notes?: string | null;
  cancelledAtUtc?: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface CreateSubscriptionRequest {
  serviceName: string;
  providerName: string;
  category: SubscriptionCategory;
  tags: string[];
  billingCycle: SubscriptionBillingCycle;
  amount: number;
  currencyCode: string;
  startDateUtc: string;
  nextDueDateUtc: string;
  reminderDaysBeforeDue: number;
  autoRenew: boolean;
  status: SubscriptionStatus;
  notes?: string | null;
}

export interface UpdateSubscriptionRequest extends CreateSubscriptionRequest {}

export interface SubscriptionAnalyticsResponse {
  trend: SubscriptionTrendPoint[];
  categoryBreakdown: SubscriptionCategoryBreakdown[];
  currentMonthlyEstimatedSpend: number;
}

export interface SubscriptionTrendPoint {
  year: number;
  month: number;
  estimatedSpend: number;
}

export interface SubscriptionCategoryBreakdown {
  category: SubscriptionCategory;
  estimatedMonthlySpend: number;
  subscriptionCount: number;
}

File: src/Flux.Data/Models/Subscription.cs
namespace Flux.Data.Models;

public sealed class Subscription
{
    public int Id { get; set; }

    public Guid OwnerUserId { get; set; }

    public string OwnerUsername { get; set; } = string.Empty;

    public string ServiceName { get; set; } = string.Empty;

    public string ProviderName { get; set; } = string.Empty;

    public SubscriptionCategory Category { get; set; }

    public string TagsCsv { get; set; } = string.Empty;

    public SubscriptionBillingCycle BillingCycle { get; set; }

    public decimal Amount { get; set; }

    public string CurrencyCode { get; set; } = "USD";

    public DateTime StartDateUtc { get; set; }

    public DateTime NextDueDateUtc { get; set; }

    public int ReminderDaysBeforeDue { get; set; }

    public bool AutoRenew { get; set; } = true;

    public SubscriptionStatus Status { get; set; } = SubscriptionStatus.Active;

    public string? Notes { get; set; }

    public DateTime? CancelledAtUtc { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

public enum SubscriptionCategory
{
    Entertainment,
    Insurance,
    Utilities,
    Mobile,
    Internet,
    Productivity,
    Health,
    Education,
    Transportation,
    Other
}

public enum SubscriptionBillingCycle
{
    Weekly,
    Monthly,
    Quarterly,
    Yearly
}

public enum SubscriptionStatus
{
    Active,
    Paused,
    Cancelled
}
